node/src/actors/chain_manager/handlers.rs

```diff
@@ -306,118 +302,234 @@ impl Handler<AddBlocks> for ChainManager {
-                            == consensus_constants.bootstrap_hash
-                            || msg.blocks[0].block_header.beacon.checkpoint
-                                > chain_beacon.checkpoint
-                        {
-                            &msg.blocks[..]
-                        } else {
-                            &msg.blocks[1..]
-                        };
+                // TODO: review if `sync_waiting_for_add_blocks_since = None` is needed
+                if self.sync_target.is_none() {
+                    log::warn!("Target Beacon is None");
+                    return;
+                }
 
-                        for block in blocks.iter() {
-                            // Update reputation before checking Proof-of-Eligibility
-                            let block_epoch = block.block_header.beacon.checkpoint;
-
-                            // Construct superblock upon consolidation of the first block that took place after
-                            // the initial checkpoint of the next superblock
-                            let next_superblock_index =
-                                self.chain_state.superblock_state.get_beacon().checkpoint + 1;
-                            if block_epoch >= next_superblock_index * superblock_period {
-                                // Construct superblocks while synchronizing but do not broadcast them
-                                // This is needed to ensure that we can validate the received superblocks later on
-                                self.construct_superblock(ctx, block_epoch)
-                                    .and_then(move |_, _act, _ctx| actix::fut::ok(()))
-                                    .wait(ctx);
-                            }
+                if msg.blocks.is_empty() {
+                    log::debug!("Received an empty AddBlocks message");
+                    self.sm_state = StateMachine::WaitingConsensus;
+                    self.initialize_from_storage(ctx);
+                    log::info!("Restored chain state from storage");
 
-                            if let Err(e) = self.process_requested_block(ctx, block.clone()) {
-                                log::error!("Error processing block: {}", e);
-                                self.initialize_from_storage(ctx);
-                                log::info!("Restored chain state from storage");
-                                batch_succeeded = false;
-                                break;
-                            }
+                    return;
+                }
 
-                            let beacon = self.get_chain_beacon();
-                            show_sync_progress(
-                                beacon,
-                                target_beacon,
-                                self.epoch_constants.unwrap(),
-                            );
+                let sync_target = self.sync_target.clone().unwrap();
```

Comment:
comment on why this cannot fail?

Reply:
Actually, the clone is no longer required! :)